## Lectura 4-3: Supuestos de la regresión lineal simple

### Interpretación probabilística

- Teniendo en cuenta los supuestos anteriores, estamos diciendo que la densidad de probabilidad (PDF) de los errores $\epsilon$ es una gaussiana con media cero y varianza constante $\sigma^2$:

 $$\text{PDF}(\epsilon_{i})=\frac{1}{\sqrt{2\pi} \sigma} \exp \left(- \frac{\epsilon_{i}^{2}}{2\sigma^2}\right)$$

-   Además todos los $\epsilon_{i}$ son IID (independientes e idénticamente distribuidos).

- Esto implica que:

 $$\text{PDF}(y_i|x_{i};\beta)=\frac{1}{\sqrt{2\pi} \sigma} \exp \left(- \frac{(y_i - h_{\beta}(x_{i}) )^{2}}{2\sigma^2}\right)$$

-  Lo que implica que la distribución de $\mathbf{y}$ dados los valores de $\mathbf{x}$, parametrizados por $\beta$ también se distribuye normalmente.

- Estimemos los parámetros del modelo ($\beta$) utilizando la estimación de máxima verosimilitud.


-   La función de verosimilitud de $\beta$ puede escribirse como:

$$\mathcal{L}(\beta) = \prod_{i=1}^{n}\frac{1}{\sqrt{2\pi} \sigma} \exp \left(- \frac{(y_i - h_{\beta}(x_{i}) )^{2}}{2\sigma^2}\right)$$

-   y el logaritmo de probabilidad $l_n(\beta)$:

$$l_n(\beta)  =  \log  \mathcal{L}(\beta) $$
$$ = \log \prod_{i=1}^{n}\frac{1}{\sqrt{2\pi} \sigma} \exp \left(- \frac{(y_i - h_{\beta}(x_{i}) )^{2}}{2\sigma^2}\right) $$
$$ = \sum_{i=1}^n  \log \frac{1}{\sqrt{2\pi} \sigma} \exp \left(- \frac{(y_i - h_{\beta}(x_{i}) )^{2}}{2\sigma^2}\right) $$
$$ = n \log\left(\frac{1}{\sqrt{2\pi} \sigma}\right) - \frac{1}{\sigma^2}\times \frac{1}{2}\sum_{i=1}^n(y_i - h_{\beta}(x_{i}) )^{2}$$

-   Después de ignorar las constantes, nos damos cuenta de que maximizar $l_n(\beta)$ es equivalente a minimizar:

$$\sum_{i=1}^n(y_i - h_{\beta}(x_{i}))^{2}$$

-   Lo que equivale a minimizar la SSE.

- Entonces, si uno estima los parámetros de $\beta$ utilizando la estimación de máxima verosimilitud se llega a los mismos resultados que haciendo la estimación por mínimos cuadrados.

- Por lo tanto, cada vez que realizamos una estimación por mínimos cuadrados, implícitamente estamos haciendo los mismos supuestos probabilísticos mencionados anteriormente (por ejemplo, normalidad, homocedasticidad).


### Ejemplo: un modelo de altura

- Vamos a trabajar con el conjunto de datos `Howell1` que contiene datos demográficos de los !Kung San del Kalahari recogidos por Nancy Howell a finales de los años sesenta.

- Los !Kung San son la población forrajera más famosa del siglo XX, en gran parte gracias a los detallados estudios cuantitativos realizados por personas como Howell.


-   Cada observación corresponde a un individuo.

- Las variables del conjunto de datos son:

- Exploremos las correlaciones lineales entre las variables

In [26]:
# Escriba su código aquí...
#read a csv called howell1.csv
import pandas as pd

df = pd.read_csv('howell1.csv', sep = ',')
df
## I have the height,weight,age,male

height     weight   age  male
0    151.765  47.825606  63.0     1
1    139.700  36.485807  63.0     0
2    136.525  31.864838  65.0     0
3    156.845  53.041914  41.0     1
4    145.415  41.276872  51.0     0
..       ...        ...   ...   ...
539  145.415  31.127751  17.0     1
540  162.560  52.163080  31.0     1
541  156.210  54.062497  21.0     0
542   71.120   8.051258   0.0     1
543  158.750  52.531624  68.0     1

[544 rows x 4 columns]

- Podemos ver que existe una correlación positiva entre `altura`  y `edad`.
- Filtremos los ejemplos no adultos porque sabemos que la altura está fuertemente correlacionada con la edad antes de la edad adulta.


In [28]:
# Escriba su código aquí...
#make a correlation between height and age
print(df.corr())
no_adultos=df[df['age']<18]
mayores=df[df['age']>=18]


          height    weight       age      male
height  1.000000  0.940822  0.683689  0.139229
weight  0.940822  1.000000  0.678335  0.155443
age     0.683689  0.678335  1.000000  0.005887
male    0.139229  0.155443  0.005887  1.000000


In [15]:
print(no_adultos.corr())

          height    weight       age      male
height  1.000000  0.944479  0.943313  0.061724
weight  0.944479  1.000000  0.917491  0.071611
age     0.943313  0.917491  1.000000 -0.006452
male    0.061724  0.071611 -0.006452  1.000000


-   Ahora bien, la edad no se correlaciona con la altura:

In [24]:
## Escriba su código aquí...
#create a linaer regression between height and age
import statsmodels.formula.api as smf
mod = smf.ols(formula='height ~ age', data=no_adultos)
res = mod.fit()
print(res.summary())


                            OLS Regression Results                            
Dep. Variable:                 height   R-squared:                       0.890
Model:                            OLS   Adj. R-squared:                  0.889
Method:                 Least Squares   F-statistic:                     1535.
Date:                Fri, 08 Sep 2023   Prob (F-statistic):           6.02e-93
Time:                        21:07:33   Log-Likelihood:                -683.84
No. Observations:                 192   AIC:                             1372.
Df Residuals:                     190   BIC:                             1378.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     73.3723      1.085     67.601      0.0

Coefficients:

    Intercept: The coefficient for the intercept is 73.3723. This is the estimated height when age is zero. However, this value may not have a practical interpretation in this context because it's unlikely to have people with an age of zero.
    Age: The coefficient for age is 4.5257. This indicates that, on average, for each unit increase in age, there is an associated increase of approximately 4.53 units in height.

P-values for Coefficients:

    The p-values associated with both the intercept and age are very close to zero (0.000), indicating that both coefficients are statistically significant. In other words, both the intercept and age have a significant effect on height.

-  Modelemos la altura en función del peso mediante una regresión lineal simple:
    
$$\text{height}(\text{weight})=\beta_0+\beta_1*\text{weight}$$


In [29]:
# Escriba su código aquí...
import statsmodels.formula.api as smf
mod = smf.ols(formula='height ~ weight', data=mayores)
res = mod.fit()
print(res.summary())

#create a linear regression between height and weight


#write the height = 113+0.9*weight



                            OLS Regression Results                            
Dep. Variable:                 height   R-squared:                       0.570
Model:                            OLS   Adj. R-squared:                  0.568
Method:                 Least Squares   F-statistic:                     463.3
Date:                Fri, 08 Sep 2023   Prob (F-statistic):           4.68e-66
Time:                        21:11:39   Log-Likelihood:                -1071.0
No. Observations:                 352   AIC:                             2146.
Df Residuals:                     350   BIC:                             2154.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    113.8794      1.911     59.589      0.0

    R-squared and Adjusted R-squared:
        R-squared (R²) measures the proportion of the variance in the dependent variable (height) that is explained by the independent variable (weight). In this case, R-squared is approximately 0.570, indicating that about 57% of the variance in height is explained by weight.

    Coefficients:
        Intercept: The coefficient for the intercept is 113.8794. This is the estimated height when weight is zero. However, this value may not have a practical interpretation in this context because it's unlikely to have people with a weight of zero. In the context of your previous question, it suggests that a person with zero weight is estimated to have a height of approximately 113.88 cm.
        Weight: The coefficient for weight is 0.9050. This indicates that, on average, for each additional unit of weight (presumably in some unit like kilograms), there is an associated increase of approximately 0.9050 units in height (presumably in some unit like centimeters).

    P-values for Coefficients:
        The p-values associated with both the intercept and weight are very close to zero (0.000), indicating that both coefficients are statistically significant. In other words, both the intercept and weight have a significant effect on height.

##### write the height = 113+0.9*weight





- Podemos ver que los coeficientes del modelo son $\beta_{0}=113,879$ y $\beta_{1}=0,905$.

- La estimación de $\beta_{0}=xxxx$, indica que una persona de peso 0 debería medir alrededor de $114$cm, lo cual sabemos ....

- Dado que $\beta_1$ es una pendiente, el valor $xxx$ puede leerse como que se espera que una persona con 1 kg más de peso sea xxx cm más alta.

- Podemos acceder directamente a los coeficientes y almacenarlos en una variable:

In [33]:
# Escriba su código aquí...



AttributeError: 'OLS' object has no attribute 'conf'

-   Podemos ver varios indicadores sobre el modelo lineal con el comando **summary**:

In [ ]:
# Escriba su código aquí...

- Podemos ver que $\beta_0$ y $\beta_1$ son ambos estadísticamente significativamente diferentes de cero.

- Vemos que el coeficiente de determinación $R^2$ tiene un valor de 0,57$ que no es muy bueno pero aceptable.

- Podemos concluir que el peso, aunque ....

- Podemos acceder al coeficiente de determinación:

In [ ]:
# Escriba su código aquí...

- También podemos acceder a los valores ajustados que son los valores predichos por mi modelo para los datos utilizados:

In [ ]:
# Escriba su código aquí...

- Podemos comprobar ahora que todas las definiciones dadas para $R^2$ son
    equivalentes.

In [ ]:
# Escriba su código aquí...

-   O calcular manualmente el error estándar del modelo $SE_{\text{model}}$:

In [ ]:
# Escriba su código aquí...



-  Supongamos ahora que conocemos el peso de dos personas !Kung pero no conocemos su altura.

- Podríamos utilizar nuestro modelo lineal para predecir la altura de estas dos personas.



In [ ]:
# Escriba su código aquí...